bytecomp/matching.ml

```diff
@@ -2881,29 +2884,66 @@ and do_compile_matching repr partial ctx arg pmh = match pmh with
         (divide_record lbl.lbl_all (normalize_pat pat))
         ctx_combine repr partial ctx pm
   | Tpat_constant cst ->
+      let names = None in
       compile_test
         (compile_match repr partial) partial
         divide_constant
-        (combine_constant pat.pat_loc arg cst partial)
+        (combine_constant names pat.pat_loc arg cst partial)
         ctx pm
-  | Tpat_construct (_, cstr, _) ->
+  | Tpat_construct (_, cstr, pats) ->
+      let names_from_type_variant cstrs =
+        let (consts, blocks) = List.fold_left
+          (fun (consts, blocks) cstr ->
+            if cstr.Types.cd_args = []
+            then (Ident.name cstr.Types.cd_id :: consts, blocks)
+            else (consts, Ident.name cstr.Types.cd_id :: blocks))
+          ([], []) cstrs in
+        Some {consts = consts |> List.rev |> Array.of_list;
+              blocks = blocks |> List.rev |> Array.of_list } in
+
+      let rec resolve_path n path =
+        match Env.find_type path pat.pat_env with
+        | {type_kind = Type_variant cstrs} ->
+          names_from_type_variant cstrs
+        | {type_kind = Type_abstract; type_manifest = Some t} ->
+          ( match (Ctype.unalias t).desc with
+            | Tconstr (pathn, _, _) ->
+              (* Format.eprintf "XXX path%d:%s path%d:%s@." n (Path.name path) (n+1) (Path.name pathn); *)
+              resolve_path (n+1) pathn
```

Comment:
where is `n` used? to avoid infinite loop?

Reply:
It's only used in the commented-out print. I have asked myself the question of whether infinite loops are possible, and one should guard against it.
Somehow, they should not, as if it's a variant, then eventually the variant definition must show up. Though I would not bet my life on it.